nxc/protocols/winrm.py

```diff
@@ -249,78 +285,80 @@ def plaintext_login(self, domain, username, password):
-            self.db.add_credential("hash", domain, self.username, nthash)
 
-            if self.admin_privs:
-                self.db.add_admin_user("hash", domain, self.username, nthash, self.host)
+            user_id = self.db.add_credential("hash", domain, self.username, nthash)
+            host_id = self.db.get_hosts(self.host)[0].id
+            self.db.add_loggedin_relation(user_id, host_id)
+            self.db.add_admin_user("hash", domain, self.username, nthash, self.host, user_id=user_id)
 
             if not self.args.local_auth:
                 add_user_bh(self.username, self.domain, self.logger, self.config)
             return True
 
         except Exception as e:
             if "with ntlm" in str(e):
-                self.logger.fail(f"{self.domain}\\{self.username}:{process_secret(nthash)}")
+                self.logger.fail(f"{self.domain}\\{self.username}:{process_secret(self.nthash)}")
             else:
-                self.logger.fail(f"{self.domain}\\{self.username}:{process_secret(nthash)} '{e}'")
+                self.logger.fail(f"{self.domain}\\{self.username}:{process_secret(self.nthash)} '{e}'")
             return False
 
-    def execute(self, payload=None, get_output=False):
-        try:
-            self.logger.debug(f"Connection: {self.conn}, and type: {type(self.conn)}")
-            r = self.conn.execute_cmd(self.args.execute, encoding=self.args.codec)
-            self.logger.success("Executed command")
-            buf = StringIO(r[0]).readlines()
-            for line in buf:
-                self.logger.highlight(line.strip())
-        except Exception as e:
-            self.logger.debug(f"Error executing command: {e}")
-            self.logger.fail("Cannot execute command, probably because user is not local admin, but running via powershell (-X) may work")
+    def execute(self, payload=None, get_output=True, shell_type="cmd"):
+        if not payload:
+            payload = self.args.execute
+
+        if self.args.no_output:
+            get_output = False
 
-    def ps_execute(self, payload=None, get_output=False):
         try:
-            r = self.conn.execute_ps(self.args.ps_execute)
-            self.logger.success("Executed command")
-            buf = StringIO(r[0]).readlines()
+            result = self.conn.execute_cmd(payload, encoding=self.args.codec) if shell_type == "cmd" else self.conn.execute_ps(payload)
+        except Exception as e:
+            # Reference: https://github.com/diyan/pywinrm/issues/275
+            if hasattr(e, "code") and e.code == 5:
+                self.logger.fail(f"Execute command failed, current user: '{self.domain}\\{self.username}' has no 'Invoke' rights to execute command (shell type: {shell_type})")
+                
+                if shell_type == "cmd":
+                    self.logger.info("Cannot execute command, probably has not invoke rights with Root WinRM listener, now auto switch to PSSession!")
```

Comment:
Replace to: `Cannot execute command via cmd, the user probably does not have invoke rights with Root WinRM listener - now switching to Powershell to attempt execution`

Reply:
Sure

Reply:
Fixed!